source/extensions/filters/http/grpc_json_transcoder/async_reflection_fetcher.cc

```diff
@@ -0,0 +1,93 @@
+#include "source/extensions/filters/http/grpc_json_transcoder/async_reflection_fetcher.h"
+
+#include "envoy/config/core/v3/grpc_service.pb.h"
+
+#include "source/common/common/lock_guard.h"
+#include "source/common/grpc/async_client_impl.h"
+
+using grpc::reflection::v1alpha::ServerReflectionRequest;
+using grpc::reflection::v1alpha::ServerReflectionResponse;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace GrpcJsonTranscoder {
+
+void AsyncReflectionFetcher::requestFileDescriptors(
+    std::function<void(Envoy::Protobuf::FileDescriptorSet)> file_descriptors_available) {
+  Thread::LockGuard guard_(receive_message_mutex_);
+  file_descriptors_available_ = file_descriptors_available;
+
+  // Register to init_manager, force the listener to wait for completion of the
+  // reflection Rpcs.
+  init_target_ = std::make_unique<Init::TargetImpl>("JsonGrpcFilter: Grpc Reflection Rpcs",
+                                                    [this]() -> void { startReflectionRpcs(); });
+  init_manager_.add(*init_target_);
+}
+
+void AsyncReflectionFetcher::startReflectionRpcs() {
+  envoy::config::core::v3::GrpcService config;
+  config.mutable_envoy_grpc()->set_cluster_name(reflection_cluster_config_.cluster_name());
+
+  const Envoy::Protobuf::MethodDescriptor* server_reflection_method =
+      Envoy::Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "grpc.reflection.v1alpha.ServerReflection.ServerReflectionInfo");
+  if (server_reflection_method == nullptr) {
+    ENVOY_LOG(error, "Failed to find MethodDescriptor for ServerReflectionInfo");
+    return;
+  }
+
+  for (size_t i = 0; i < services_.size(); i++) {
+    std::string service = services_[i];
+    Envoy::Grpc::AsyncClient<ServerReflectionRequest, ServerReflectionResponse>& async_client =
+        async_clients_[i];
+
+    AsyncReflectionFetcherCallbacks* callbacks = new AsyncReflectionFetcherCallbacks(
```

Comment:
Where is the `callbacks` freed? ???

Reply:
Sorry I missed this earlier.

This is a good catch, that they are not freed!

However, it does leave an interesting question:
When do you think it's safe to free them in case of failure?
For success, I think we can free them after we set init_target_->ready().
For failure, there are approaches I can currently think of:
a. We can keep a count of any kind of return, success or failure, and free when that count is reached.
b. We can record that an error happened, and erase from `remaining_callbacks_` just like in success. 
When `remaining_callbacks_` reach zero, we release regardless of success or failure, but only call ready() when the error flag is not set.

Thoughts on which is preferable, or do you have other ideas for deciding when to free?

Reply:
Actually, you do have `callbacks_` to hold the unique_ptr for them,  they are freed eventually when `AsyncReflectionFetcher` is freed.

Just a suggestion,  `remaining_callbacks_` can be just a integer to keep track of number of outstanding requests.

Reply:
I made it a set because I was not sure if internal retries in the Envoy gRPC library could cause duplicate responses.

Do you happen to know if it duplicate responses are guaranteed to be handled and de-duped?

Reply:
All envoy async callbacks are guaranteed to be called once.  Not need to worry about been called twice.  In your code, if onCloseStream() is called on network failure,  it will not call your callback.  But that is fine since we decide not to "ready" envoy in failed cases.